programs/clearing_house/src/math/lp.rs

```diff
@@ -0,0 +1,445 @@
+                total_lp_tokens,
+            )?,
+            amm.peg_multiplier,
+        )?;
+
+        let min_qaa = amm.minimum_quote_asset_trade_size;
+        let min_baa = amm.minimum_base_asset_trade_size;
+
+        if base_asset_amount.unsigned_abs() >= min_baa && quote_asset_amount >= min_qaa {
+            market_quote_asset_amount = quote_asset_amount;
+            market_base_asset_amount = base_asset_amount;
+        } else {
+            // no market position bc too small so give them negative upnl
+            // similar to closing their small position
+            // TODO: decide what this should be
+            unsettled_pnl = cast_to_i128(min_qaa)?
+                .checked_mul(-1)
+                .ok_or_else(math_error!())?;
+        }
+    }
+
+    let lp_metrics = LPMetrics {
+        fee_payment,
+        funding_payment,
+        base_asset_amount: market_base_asset_amount,
+        quote_asset_amount: market_quote_asset_amount,
+        unsettled_pnl,
+    };
+
+    Ok(lp_metrics)
+}
+
+pub fn calculate_swap_quote_reserve_delta(
+    amm: &AMM,
+    base_asset_amount: i128,
+) -> ClearingHouseResult<u128> {
+    let swap_direction = match base_asset_amount > 0 {
+        true => SwapDirection::Remove,
+        false => SwapDirection::Add,
+    };
+
+    let (new_quote_asset_reserve, _) = calculate_swap_output(
+        base_asset_amount.unsigned_abs(),
+        amm.base_asset_reserve,
+        swap_direction,
+        amm.sqrt_k,
+    )?;
+
+    // avoid overflow - note: sign doesnt matter
+    let quote_asset_reserve_output = if new_quote_asset_reserve > amm.quote_asset_reserve {
+        new_quote_asset_reserve
+            .checked_sub(amm.quote_asset_reserve)
+            .ok_or_else(math_error!())?
+    } else {
+        amm.quote_asset_reserve
+            .checked_sub(new_quote_asset_reserve)
+            .ok_or_else(math_error!())?
+    };
+
+    Ok(quote_asset_reserve_output)
+}
+
+pub fn get_lp_market_position_margin(
+    lp_position: &MarketPosition,
+    amm: &AMM,
+) -> ClearingHouseResult<(MarketPosition, u128)> {
+    let total_lp_tokens = amm.sqrt_k;
+    let lp_tokens = lp_position.lp_tokens;
+
+    let mut market_position = *lp_position; // clone bc its only temporary
+    let lp_metrics = get_lp_metrics(&market_position, lp_tokens, amm)?;
+
+    // update pnl payments
+    market_position.unsettled_pnl = lp_position
+        .unsettled_pnl
+        .checked_add(lp_metrics.fee_payment)
+        .ok_or_else(math_error!())?
+        .checked_add(lp_metrics.funding_payment)
+        .ok_or_else(math_error!())?;
+
+    // worse case market position
+    // max ask: (sqrtk*1.4142 - base asset reserves) * lp share
+    // max bid: (base asset reserves - sqrtk/1.4142) * lp share
+
+    // TODO: is there a cleaner way to do this? -- maybe make it a constant?
+    let percision: f64 = 10_000_000_000_000.0; // amm percision as float
+    let sqrt_2 = (2_f64.sqrt() * percision).round() as u128;
```

Comment:
this is probably spends a lot of compute, you can just approximate it instead by multiply by `14142` and dividing by `10000`